storage/pebble/blocks_provider.go

```diff
@@ -0,0 +1,130 @@
+package pebble
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-evm-gateway/storage"
+	"github.com/onflow/flow-go/fvm/evm/events"
+	evmTypes "github.com/onflow/flow-go/fvm/evm/types"
+	flowGo "github.com/onflow/flow-go/model/flow"
+	gethCommon "github.com/onflow/go-ethereum/common"
+	"github.com/onflow/go-ethereum/eth/tracers"
+)
+
+type ResultsCollector struct{}
+
+func (s *ResultsCollector) StorageRegisterUpdates() map[flowGo.RegisterID]flowGo.RegisterValue {
+	return map[flowGo.RegisterID]flowGo.RegisterValue{}
+}
+
+var _ evmTypes.ReplayResultCollector = (*ResultsCollector)(nil)
+
+type BlocksProvider struct {
+	blocks             storage.BlockIndexer
+	chainID            flowGo.ChainID
+	tracer             *tracers.Tracer
+	latestBlockPayload *events.BlockEventPayload
+}
+
+var _ evmTypes.BlockSnapshotProvider = (*BlocksProvider)(nil)
+var _ evmTypes.BlockSnapshot = (*BlocksProvider)(nil)
+
+func NewBlocksProvider(
+	blocks storage.BlockIndexer,
+	chainID flowGo.ChainID,
+	tracer *tracers.Tracer,
+) *BlocksProvider {
+	return &BlocksProvider{
+		blocks:  blocks,
+		chainID: chainID,
+		tracer:  tracer,
+	}
+}
+
+func (bp *BlocksProvider) OnBlockReceived(blockEvent *events.BlockEventPayload) error {
+	if bp.latestBlockPayload != nil {
```

Comment:
instead of using `latestBlockPayload` evmTypes.BlockSnapshot should be implemented by a separate object.

something like 
```
type blockSnapshot {
  BlocksProvider
  height uint64
}
```

the block context would then look like:

```
func (bs *blockSnapshot) BlockContext() (evmTypes.BlockContext, error) {
   return bs.BlocksProvider.BlockContext(bs.height)

}
```

while GetSnapshotAt would look like:
```
func (bp *BlocksProvider) GetSnapshotAt(height uint64) (
	evmTypes.BlockSnapshot,
	error,
) {
	return *blockSnapshot{
           BlocksProvider: bp,
           height: height
        }, nil
}
```

Reply:
Good point :+1: The `BlocksProvider` is rather counter-intuitive, with implementing both interfaces, so I have moved the `BlockSnapshot` interface to a dedicated object. See: https://github.com/onflow/flow-evm-gateway/pull/632/commits/d9890bfd4a476125432c7a4deb7f28cdea5266ff#diff-8eafa2e76b08bd95e6922ee08584718a271ae9f9bf7d1d18bd6c959ee6edb7a3R23-R28